### ISSUE TITLE
test(metalgemm): preserve native Q8 GEMV arithmetic in P4 graph projections

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12920]
Parent: #12349
Coordinates with: #12853
Coordinates with: #12914
Coordinates with: #12919

## Parent context
#12919 physical native M3 Pro observation matches all1024 codes/all32 scale bits and H=D for token7 layer0 mixed QKV. H and D still differ from native serial S: bf7e14ef versus bf7e14f0 at channel0. Full strict parity remains RED. This child isolates the existing Q8 projection arithmetic route while retaining P4 graph batching.

## Current state
Foundation is frozen #12919: full1702 manifest36122cb1852de431cdd7e04f943cad8468c5dee4dc7151479426124850307d51, q4k f7c3bdc617c66dda92acf0e5614ff0495502c5c6edff23520396eff3354a54df, strict testd682fe65cde687886a0368f3c6bea8eed2e6b5d03343b3b7f165212010742b13. Physical receipt4831985b05b2636cb0546f8c46180a48ab80a402b6f0cdb929743d0d7c2d56c6 records peer overlap and no performance claim.

Serial q8_gemv uses int32 dot products for each32-code block, multiplies by weight/activation scales, accumulates per lane with block stride32, then simd_sum. Existing q8_gemm dequantizes each code into float before element multiply-accumulate. Same quantized input therefore need not produce identical float bits.

## For / Problem / Today / Better because
For exact Metal MTP verification, retain the native serial projection arithmetic in a single P4 graph dispatch. Today graph GEMM and serial GEMV use different accumulation expressions. Extending the existing GEMV's grid to output-row/token coordinates preserves the arithmetic source while batching four tokens without a production CPU token loop or extra command buffer.

## Why this is next
Direct #12919 intermediates remove code/scale drift from the captured H-versus-S comparison. Reusing the already-witnessed native serial kernel is narrower than inventing another projection kernel or changing the exact oracle.

## Prior-art/SOTA source check
Observed2026-09-12. Reused sota-check matching metal-quant-gemm matrix row (captured JSON /tmp/fak-q8-p4-proposal/sota-row.json); `fak sota internal/metalgemm/q8.m` resolves that row. Matrix primary llama.cpp Metal/MLX was inspected at3057bb66c86c46d5781e50e85462a760ba7d1feb in the #12919 prior-art receipt, including MIT license94f29bbed6a22c35b992c5c6ebf0e7c92f13b836b90f36f461c9cf2f0f1d010d and per-library source compilation. No upstream kernel code is copied or engine substituted. Route stay-minimal: reuse FAK's existing serial arithmetic, with the stronger exact-bit oracle rather than the matrix's cosine floor.

Apple's dispatchThreadgroups API takes independent grid and threadgroup dimensions: https://developer.apple.com/documentation/metal/mtlcomputecommandencoder/dispatchthreadgroups%28_%3Athreadsperthreadgroup%3A%29?language=objc . Local q8.m already uses explicit grid/threadgroup arguments. This establishes the host dispatch seam, not arithmetic parity. Kernel threadgroup position becomes uint3; only x/y used, z grid extent1.

Portfolio WATCH until physical qualification for P4/M3Pro. General P batching, new GEMM algorithms, new shader entry points and performance promotion are EXCLUDE from this leaf. Refresh on witness or compiler/device changes. Existing #12919 licensing/docs omissions remain declared; no broad upstream completeness claim.

## Working spine
Add P at existing q8_gemv buffer7. Read output row from position.x and token from position.y, check both bounds, then advance X by token*nblk*32, XD by token*nblk and Y by token*out. Keep weight offsets, full int32 block-dot loop, float scale accumulation, lane block-stride and simd_sum/output expression verbatim.

Bind P=1 at all four existing native GEMV encoder sites. In both mg_q8_graph_encode and mg_q8_graph_encode_from select that existing pipeline only when P==4 and encode one dispatch with grid(out,P,1), threads(32,1,1). All other P values retain existing psoQ8Gemm and tiled dispatch. Graph buffers retain token-major layout and same allocation sizes. No Go/bridge API, graph receipt, new shader entrypoint, intermediate fence/readback, CPU token loop or fallback is added.

## Exact P1 binding and layout audit
Base q8.m SHA2569effcb3898da59821265fbd70a5317e1767bcdbbbcd0ed246c15ceb31368beb8. Repository-wide symbol search finds exactly four existing psoQ8Gemv encoder sites, all in this file:
- mg_q8_encode_gemv_group, line350: bind P1 once on its encoder before the weight loop; retain per-weight output buffer offset.
- mg_q8_gemv, line388: bind P1 before standalone dispatch.
- mg_q8_gemv_group, line430: bind P1 once before its weight loop, preserving yoff packing.
- mg_issue8833_q8_encode_gemv, line464: bind P1 in the caller-owned helper, preserving ownership and no internal commit/read.

Graph entry points at614/623 already bind P at buffer7; extend their pipeline selection and dispatch symmetrically. P4 token-major code/scale/output offsets are bounded by existing P*in/P*nblk/P*out allocation. Long arithmetic is used for offsets. P1 has token0, so these offsets are zero. Non-P4 graph P1/P2/P32 retains existing GEMM selection. Changing shader parameters can still affect compiled code, so P1 behavior must be witnessed rather than presumed.

## Proposed diff
Unapplied combined two-file proposal `/tmp/fak-q8-p4-proposal/proposed-unapplied.patch`, SHA256caa03075f56eab9cb899d3936c9e892289e746e19898a5f6aaf0ce4bb9e92029. Production-only portion q8-unapplied.patch132eb4f263417e02a4a4f42ebbfb5c18c9d725e5ae3ef7a2d6d0f5a91fe89ccf. No repository source changed.

One focused existing-file regression is included in graph_darwin_test.go (base39865698a0af6385b425d6f45a2bb1ef62b386996b1c9f9dcf8d07fae28702f8): TestProjectionGraphQ8P4MatchesSerialGEMV, one fixture P4/in1280/out37. Forty blocks exercise lane stride;37 outputs exercise a tail; distinct token-major inputs use exactly representable known quantization scales. Compare every row/column of both EncodeQ8 and QuantizeQ8+EncodeQ8From to native P1 serial GEMV, bitwise, with finite/nonempty references and one packed graph receipt. No extra generic suite or tolerance. Reference-only token loop is in the test, not production. Parent #12919 H-versus-S failure is the captured pre-change reproducer; new focused test has not run yet.

## Initialization and failure behavior
Reuse existing psoQ8Gemv and q8_init options:nil, with existing required-pipeline gate. New binding is supplied to every existing native caller; no stale/missing buffer7 expected. Existing graph input/weight/init validation remains unchanged. No new recovery/fallback path; if init or dispatch execution is unavailable, witness is unproven, not accepted. No changes to ownership or global reset behavior.

## Core through-line
Reuse one existing arithmetic sequence in a2D P4 dispatch, independently qualify host bindings/layout and exact outputs, then keep the original full strict oracle authoritative.

## Scope
```routing
lane: metalgemm
paths: ["internal/metalgemm/q8.m", "internal/metalgemm/graph_darwin_test.go"]
expected_steps: 3
```

## Likely files
- internal/metalgemm/q8.m
- internal/metalgemm/graph_darwin_test.go

## Scope class
Work unit: leaf. One production file and one focused regression in an existing test file. Proposal review before edits; canonical admission/lease and managed worker scope before applying.

## Definition of done
- [ ] Independent exact diff review covers every P1 binding, P4 layout/selection and unchanged arithmetic.
- [ ] Applied candidate and unchanged parent test independently reviewed and frozen.
- [ ] Separately granted physical validation captures focused graph parity, relevant existing P1 regressions and original strict result.
- [ ] Independent evidence review records fixture-scoped disposition; no presumed full parity or performance claim.

## Gold-plating boundary
No new kernel/algorithm, broad P routing, CPU loop in production, test tolerance, strict test edits, graph API/receipt change, fallback engine, new compile flags or performance claim. Maintain #12919 isolated quantizer compilation. No edit/run before review.

## Acceptance gate
Require native physical execution, no skips. New focused P4 test must pass exact all-row parity for both graph entrypoints. Existing TestMetalQ8GemvMatchesCPU and TestMetalQ8GemvGroupMatchesSingle cover standalone/group P1; TestMixedQ4KQ8Observation covers mixed-group helper; TestIssue8833Q8CallerOwnedBridge covers caller-owned helper. Existing TestProjectionGraphMixedQuantizedSingleFence checks unchanged non-P4 graph path. Existing TestMetalMTPP4StrictSerialParity must report all1024 codes/all32 scale bits, H/D/S, lifecycle/accounting, and full strict outcome separately. H=D=S is the local projection target; full strict parity is measured without presumption. Existing P1 test tolerances are not strengthened or weakened in this proposal.

## Done condition
Captured exact projection and P1 binding evidence with an independent disposition. Parent #12349 remains open unless all original acceptance conditions actually pass; no arithmetic/source equivalence claim alone suffices.

## Witness
Proposed serialized two-package invocation, subject to explicit budget review:
env -u FAK_METAL_MPS CGO_ENABLED=1 GOMAXPROCS=2 GOFLAGS=-p=1 GOCACHE=/tmp/fak-12853-reduction-run/cache GOTMPDIR=/tmp/fak-12853-reduction-run/tmp go test ./internal/metalgemm ./internal/model -run '^(TestProjectionGraphQ8P4MatchesSerialGEMV|TestProjectionGraphMixedQuantizedSingleFence|TestMetalQ8GemvMatchesCPU|TestMetalQ8GemvGroupMatchesSingle|TestMixedQ4KQ8Observation|TestIssue8833Q8CallerOwnedBridge|TestMetalMTPP4StrictSerialParity)$' -count=1 -timeout=60s -json

## Budget
No CPU/HIL now. Proposed one validation invocation, packages serialized by -p1, timeout60 per package, owned group TERM230/KILL240/census270/report300 total. This is broader than the previous single-test grant and requires its own explicit review/grant. Existing largest P1 case5120x5120 is included for standalone coverage; no benchmark. Source/owner drift cancels; no auto-reprepare/retry. A failing package does not suppress inspection of other actually executed test results. Record competing work and exclude timing claims.

## Duplicate search
Open/closed REST search Q8 P4 found #12914/#12919/#12853/#12349 and adjacent serving/CPU work; none isolates this same-GEMV P4 graph route. Repository-wide psoQ8Gemv and graph-entrypoint search confirms the exact two production graph seams and four P1 bindings.

## Problem frame
Centrality: Core
Portfolio tier: 2 serving
P1: advanced - exact native projection qualification.
P2: preserved - no performance promotion.
P3: preserved - existing kernel and bounded P4 routing.
P4: advanced - direct physical exact oracle plus P1 regressions.

## Work estimate
Estimate: 1 point.

## Overall completion contribution
Contribution: 1/4 points within #12349; planning only.

## Completion standard
experiment

## Closure binding
Witnessed explicit-path commit cites child and (fak metalgemm) with Prior-art trailer naming the existing native Q8 block-dot route and Apple grid-dispatch API, or independently record a non-landing experiment disposition. No full-parent closure without the original gates.

**Comment by anthony-chaudhary** [comment]
Approved #12920 two-file proposal applied exactly and frozen for final source-identity review; no build/test/HIL/commit.

Managed worker: fak-worker-wt-metalgemm-be633f987f73. Artifacts: _scratch/issue-12920-source/{candidate.patch,source.sha256,receipt.json,before/,approved-proposal.md}.

- Patch caa03075f56eab9cb899d3936c9e892289e746e19898a5f6aaf0ce4bb9e92029 matches approved proposal exactly.
- Full1702 source manifest ab896f4bd0b5051b607da2d7f578cef57fd9b5c7555e53af863e6f2be8fdccf8.
- q8.m d4794900b23c817ecb2b251328a48d3ddf1a2e6ee930b36606cd6d3323e4dd9a; graph_darwin_test.go cd98a7c9281a591411432587fb6bd37685a1a55df8f1a2a09385269808087438.
- Original strict test d682fe65cde687886a0368f3c6bea8eed2e6b5d03343b3b7f165212010742b13 and #12919 quantizer q4k.m f7c3bdc617c66dda92acf0e5614ff0495502c5c6edff23520396eff3354a54df remain unchanged.
- Programmatic checks confirm only the two authored paths differ, original inner arithmetic/simd_sum/output remain verbatim, four P1 bindings and both P4-only graph branches are present. Apply/diff checks pass; gofmt reports no changes.

Canonical lease released. Next: final source-identity review before a separately granted single validation invocation, two serialized packages, test60 per package and owned-group230/240/270/300 envelope. No parity/performance claim or retry.

**Comment by anthony-chaudhary** [comment]
One granted functional/bitwise invocation completed for #12920: actual executor8090 exit1, native Apple M3 Pro, two serialized packages, seven named tests run, zero skips. Six pass; original full strict parity remains RED.

PASS:
- TestProjectionGraphQ8P4MatchesSerialGEMV: all four rows/all37 outputs at in1280 (40 blocks), both graph entrypoints bitwise equal to native P1.
- TestProjectionGraphMixedQuantizedSingleFence (non-P4 graph).
- TestIssue8833Q8CallerOwnedBridge.
- TestMixedQ4KQ8Observation.
- TestMetalQ8GemvMatchesCPU (including5120x5120).
- TestMetalQ8GemvGroupMatchesSingle.

Original TestMetalMTPP4StrictSerialParity executes and remains FAIL. Its direct intermediate observations now show all1024 codes and all32 scale bit patterns exact, all four norm rows exact, and H=D=S for token7 layer0 mixed QKV, common SHA256cf28fa88ba3dbc73633df3c321ed8127753dfb02f84d31f300c02b0786f7deb0. Lifecycle checks and separate1152-byte diagnostic accounting remain valid; production20480read/6272upload/four encoders unchanged.

Remaining strict mismatch logs: verification logits rows1 and3, committed boundary, layer3 KRaw/KRoPE/V, and target-hidden history at commit/continuation. Earlier layer0–2 GDN and continuation-logit mismatch logs no longer appear. This does not establish full model parity; frozen source does not contain the separately owned hidden-capture overlays.

All1702 source entries remain unchanged at fullmanifestab896f4bd0b5051b607da2d7f578cef57fd9b5c7555e53af863e6f2be8fdccf8. Owned PID/PGID69547 and recorded descendants absent at terminal; no TERM/KILL/timeout/device error. Lease released. Resident80431 retains its original incarnation and was untouched. Functional contention policy allowed concurrent CPU builders and possiblyactive resident80431; elapsed204.527s is operational only, with no performance inference.

Preserved artifacts in managed fak-worker-wt-metalgemm-be633f987f73/_scratch/issue-12920-hil/{receipt.json,test.jsonl,test.log,supervision.json,before-check.json,after-check.json}. Receipt SHA256304d3b2a359ea48ab1a24cc398d07acaeb4276f782971d2525127677cfc062c6; rawJSONL4fea76bdc9fb293cb301d240a8b22540e89f905df389e2eb4ee98ca9b8ae899e. Source remains frozen/unlanded, one invocation only, no retry.

Next: independent raw-evidence review, then scope the remaining layer3 attention/logit discrepancy separately from the already tracked hidden-history work. Parent #12349 remains open; no new edit or run is implied.
